the-collab-lab/tcl-19-smart-shopping-list
# Replace window.alert and window.confirm with custom modal

## Description
- This code replaces `window.confirm` and `window.alert` with custom modal on all pages.
- Moved all utility functions to their own directory.
- Fixed bug of reading from `firestore` before user completes entering an existing token.
- Fixed bug of rejecting existing token when entered with leading or trailing white space.
- Modals are accessible - aria labels have been added, along with the correct focus and tab order when the modal is open and closed.


## Related Issue


Closes #30 
Closes #37 
Closes #41 
Closes #43 
Closes #40 
## Acceptance Criteria


- `window.confirm` and `window.alert` should be replaced with custom modal
- All utility functions must be in their own directory
- Check if token entered by a user exists only after the user completes entering it
- App shouldn't reject existing token entered with leading or trailing spaces.
- Aria attributes included within modals where appropriate
- Modal tab sequence is correct when modal is displayed
- When modal is open, the focus is correct (on the OK or similar button)
- When the modal closes, the focus returns to the element that invoked the modal
## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|  ✓ | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
| ✓  | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before


![image](https://user-images.githubusercontent.com/52580190/109156485-f2a02480-7781-11eb-9056-c7c280ca16db.png)


### After


![image](https://user-images.githubusercontent.com/52580190/109156198-9c32e600-7781-11eb-925c-83f66b811a95.png)


## Testing Steps / QA Criteria


- Pull this branch from gitHub using `git pull origin jm-cu-modal-component`
- Start the app on `localhost` using `npm run start`
- Try deleting an item from shopping list. There must be a prompt on a custom modal
- Try adding an item to shopping list before entering item name. A custom modal should be displayed with a message.
- Navigate to home page. Enter a non-existent token. You should see a modal with a message.
